### PR TITLE
Improved shader support(HLSL) for Unity plugin

### PR DESCRIPTION
**Pull request by krasnotsvetov**
That PR adds shader support based on shader support for ReSharper C++
https://blog.jetbrains.com/rscpp/resharper-cpp-2020-1-eap-hlsl/

### CODE REVIEW

File: resharper/resharper-unity/src/HlslSupport/CgIncludeDirectoryTracker.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                {
+                    unityReferencesTracker.HasUnityReference.AdviseOnce(lifetime, _ =>
+                    {
+                        if (solutionCaches.PersistentProperties.TryGetValue(CG_FOLDER_PATH, out var result))
+                        {
+                            var oldPath = FileSystemPath.TryParse(result, FileSystemPathInternStrategy.INTERN);
+                            var newPath = GetCgIncludeFolderPath(unityVersion);
+                            if (!oldPath.Equals(newPath))
+                            {
+                                cppGlobalCache.IsInitialUpdateFinished.Change.Advise(lifetime, v =>
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/HlslSupport/CgIncludeDirectoryTracker.cs:35]: Does this have to wait until the update is finished before invalidating? Can we cancel the update and invalidate early?

File: resharper/resharper-unity/src/HlslSupport/CgIncludeDirectoryTracker.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                                {
+                                    if (v.HasNew && v.New)
+                                    {
+                                        shellLocks.Tasks.StartNew(lifetime, Scheduling.MainGuard, TaskPriority.High, () =>
+                                        {
+                                            logger.Verbose("Dropping C++ cache, because Unity version is changed");
+                                            cppGlobalCache.ResetCache();
+                                            solutionCaches.PersistentProperties[CG_FOLDER_PATH] = newPath.FullPath;
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/HlslSupport/CgIncludeDirectoryTracker.cs:43]: This persistent property isn't used anywhere else. Is it just a marker to know to rebuild the cache?

File: resharper/resharper-unity/src/HlslSupport/ShaderLabCppFileLocationHelper.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            IPsiSourceFile sourceFile)
+        {
+            var lexer = new ShaderLabLexerGenerated(sourceFile.Document.Buffer);
+            lexer.Start();
+            bool prevCGInclude = false;
+            while (lexer.TokenType != null)
+            {
+                if (lexer.TokenType == ShaderLabTokenType.CG_INCLUDE)
+                    prevCGInclude = true;
+                else if (lexer.TokenType == ShaderLabTokenType.CG_CONTENT)
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/HlslSupport/ShaderLabCppFileLocationHelper.cs:21]: Should also handle `HLSLPROGRAM`, `HLSLINCLUDE`, `GLSLPROGRAM` and `GLSLINCLUDE`.

File: resharper/resharper-unity/src/HlslSupport/ShaderLabCppFileLocationTracker.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            var cgIncludeFolder = CgIncludeDirectoryTracker.GetCgIncludeFolderPath(myUnityVersion);        
+            if (!cgIncludeFolder.ExistsDirectory)
+                yield break;
+            
+            var range = cppFileLocation.RootRange;
+            Assertion.Assert(range.IsValid, "range.IsValid");
+            var program = sourceFile.GetDominantPsiFile<ShaderLabLanguage>()?.GetContainingNodeAt<ICgContent>(new TreeOffset(range.StartOffset));
+            Assertion.Assert(program != null, "program != null");
+            
+            if (program?.Parent?.FirstChild is ICgProgramBlock)
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/HlslSupport/ShaderLabCppFileLocationTracker.cs:83]: Probably also check for `ICgIncludeBlock`? And what about HLSL and GLSL - are these files included for those types of shaders?

File: resharper/resharper-unity/src/HlslSupport/UnityShaderModuleProvider.cs [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    //         if (!(element is IClass @class))
+    //             return false;
+    //
+    //         if (!(@class.Module is UE4BuildAndTargetFilesModule projectPsiModule))
+    //             return false;
+    //
+    //         flags = ImplicitUseKindFlags.Access | ImplicitUseKindFlags.InstantiatedNoFixedConstructorSignature;
+    //         return true;
+    //     }
+    // }
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/HlslSupport/UnityShaderModuleProvider.cs:250]: Let's delete this

File: resharper/resharper-unity/src/ShaderLab/Feature/Services/Formatting/ShaderLabReformatCode.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      if (!profile.GetSetting(ourDescriptor)) return;
+
+      var psiServices = sourceFile.GetPsiServices();
+      IShaderLabFile[] files;
+      using (new ReleaseLockCookie(psiServices.Locks, LockKind.FullWrite))
+      {
+        psiServices.Locks.AssertReadAccessAllowed();
+        files = sourceFile.GetPsiFiles<ShaderLabLanguage>().Cast<IShaderLabFile>().ToArray();
+      }
+      using (progressIndicator.SafeTotal("Reformat C/C++", files.Length))
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/ShaderLab/Feature/Services/Formatting/ShaderLabReformatCode.cs:66]: "Reformat ShaderLab"

File: resharper/resharper-unity/test/src/CSharp/Feature/Services/CodeCompletion/VersionSpecificCompletionListTest.cs [hunk 7]
```diff
@@ -35,12 +35,14 @@ public abstract class VersionSpecificCompletionListTest : CodeCompletionTestBase
...
     // These two tests have different parameters in different versions, make sure
     // we using the right version
+    [Ignore("TODO: krasnotsvetov")]
     [TestUnity(UnityVersion.Unity54)]
     public class Unity54CompletionListTest : VersionSpecificCompletionListTest
     {
         [Test] public void OnParticleTriggerWithOneArg54() { DoNamedTest(); }
     }
 
+    [Ignore("TODO: krasnotsvetov")]
```
Comment (citizenmatt) [inline resharper/resharper-unity/test/src/CSharp/Feature/Services/CodeCompletion/VersionSpecificCompletionListTest.cs:45]: Can these be un-ignored now?

File: rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/ideaInterop/fileTypes/cg/CgFileType.kt [hunk 8]
```diff
@@ -4,7 +4,7 @@ import com.jetbrains.rider.ideaInterop.fileTypes.RiderLanguageFileTypeBase
 import icons.UnityIcons
 
 object CgFileType : RiderLanguageFileTypeBase(CgLanguage) {
-    override fun getDefaultExtension() = "cginc"
+    override fun getDefaultExtension() = "glsl"
```
Comment (citizenmatt) [inline rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/ideaInterop/fileTypes/cg/CgFileType.kt:7]: This is a bit weird that the "Cg" file type is for `.glsl` files. I understood that they are different file formats.

File: rider/src/main/resources/META-INF/plugin.xml [hunk 9]
```diff
@@ -47,11 +47,16 @@
...
+    <lang.documentationProvider language="ShaderLab" implementationClass="com.jetbrains.rdclient.quickDoc.FrontendDocumentationProvider"/>
+    <backend.typedHandler language="ShaderLab" implementationClass="com.jetbrains.rider.cpp.fileType.CppTypedHandler"/>
+    <langCodeStyleSettingsProvider implementation="com.jetbrains.rider.plugins.unity.ideaInterop.fileTypes.shaderLab.settings.ShaderLabStyleSettingsProvider"/>
+
+    <fileType name="C++" extensions="cg;cginc;hlslinc;compute" />
 
     <!-- Cg support -->
     <!-- Don't forget to update CgProjectFileType list on the backend -->
-    <fileType name="Cg" fieldName="INSTANCE" implementationClass="com.jetbrains.rider.plugins.unity.ideaInterop.fileTypes.cg.CgFileType" language="Cg" extensions="cginc;compute;hlsl;glsl;hlslinc;glslinc"/>
+    <fileType name="Cg" fieldName="INSTANCE" implementationClass="com.jetbrains.rider.plugins.unity.ideaInterop.fileTypes.cg.CgFileType" language="Cg" extensions="glsl;glslinc"/>
```
Comment (citizenmatt) [inline rider/src/main/resources/META-INF/plugin.xml:59]: Should this be renamed from "Cg" to "GLSL"?

File: resharper/resharper-unity/src/HlslSupport/CgIncludeDirectoryTracker.cs (hunk 2 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/HlslSupport/CgIncludeDirectoryTracker.cs:43]: Yes

File: resharper/resharper-unity/src/HlslSupport/ShaderLabCppFileLocationHelper.cs (hunk 3 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/HlslSupport/ShaderLabCppFileLocationHelper.cs:21]: It is a bug, thank you!

File: resharper/resharper-unity/src/HlslSupport/ShaderLabCppFileLocationTracker.cs (hunk 4 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/HlslSupport/ShaderLabCppFileLocationTracker.cs:83]: we do not have support for GLSL right now, HLSL does not include those files, ICgIncludeBlock is missed, thank you!

File: resharper/resharper-unity/src/HlslSupport/ShaderLabInclusionContextProvider.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            };
+
+            var cgIncludeFolder =
+                CgIncludeDirectoryTracker.GetCgIncludeFolderPath(cache.Solution.GetComponent<UnityVersion>());
+            if (!cgIncludeFolder.IsEmpty)
+            {
+                properties.IncludePaths.Add(cgIncludeFolder);
+            }
+
+            var shaderCache = cache.Solution.GetComponent<ShaderLabCppFileLocationTracker>();
```
[the remark this answers is not in the document]
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/HlslSupport/ShaderLabInclusionContextProvider.cs:40]: oops :)

File: rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/ideaInterop/fileTypes/cg/CgFileType.kt (hunk 8 above)
Comment (krasnotsvetov) [reply rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/ideaInterop/fileTypes/cg/CgFileType.kt:7]: I hope, that we will drop CG soon, it is out-of-date shader support, I have decided to not change anything in old shader support

File: rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/ideaInterop/fileTypes/shaderLab/ShaderLabSyntaxHighlighterProvider.kt [hunk 11]
```diff
@@ -6,15 +6,16 @@ import com.intellij.openapi.fileTypes.SyntaxHighlighterFactory
 import com.intellij.openapi.fileTypes.SyntaxHighlighterProvider
 import com.intellij.openapi.project.Project
 import com.intellij.openapi.vfs.VirtualFile
+import com.jetbrains.rider.cpp.fileType.CppSyntaxHighlighter
 
 class ShaderLabSyntaxHighlighterProvider : SyntaxHighlighterProvider, SyntaxHighlighterFactory() {
     override fun getSyntaxHighlighter(project: Project?, file: VirtualFile?): SyntaxHighlighter {
-        return ShaderLabSyntaxHighlighter()
+        return CppSyntaxHighlighter()
```
[the remark this answers is not in the document]
Comment (krasnotsvetov) [reply rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/ideaInterop/fileTypes/shaderLab/ShaderLabSyntaxHighlighterProvider.kt:13]: I do not find any issues, thus while I have a lot of other task for shader integration, I have decided to use cpp lexer, but I am going to create mixed lexer soon

[the remark this answers is not in the document]
Comment (krasnotsvetov) [reply rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/ideaInterop/fileTypes/shaderLab/ShaderLabSyntaxHighlighterProvider.kt:13]: https://youtrack.jetbrains.com/issue/RIDER-47694

File: resharper/resharper-unity/test/src/CSharp/Feature/Services/CodeCompletion/VersionSpecificCompletionListTest.cs (hunk 7 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/test/src/CSharp/Feature/Services/CodeCompletion/VersionSpecificCompletionListTest.cs:45]: Right now, there is problem with that test, I will investigate it before release

File: resharper/resharper-unity/src/ShaderLab/Feature/Services/TypingAssists/ShaderLabTypingAssist.cs [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+      typingAssistManager.AddTypingHandler(lifetime, '"', this, c => WrapCppAction(c, quoteHandler.HandleQuoteTyped),
+        IsTypingSmartParenthesisHandlerAvailable);
+      typingAssistManager.AddTypingHandler(lifetime, '\'', this, c => WrapCppAction(c, quoteHandler.HandleQuoteTyped),
+        IsTypingSmartParenthesisHandlerAvailable);
+      typingAssistManager.AddTypingHandler(lifetime, ';', this, c => WrapCppAction(c, braceHandler.HandleSemicolonTyped),
+        IsCorrectCommonTyposAvailable);
+      typingAssistManager.AddTypingHandler(lifetime, ':', this, c => WrapCppAction(c, braceHandler.HandleColonTyped),
+        IsTypingHandlerAvailable);
+      typingAssistManager.AddTypingHandler(lifetime, '*', this, c => WrapCppAction(c, braceHandler.HandleStarTyped),
+        IsCorrectCommonTyposAvailable);
```
[the remark this answers is not in the document]
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/ShaderLab/Feature/Services/TypingAssists/ShaderLabTypingAssist.cs:74]: I am not sure that we need typing assit for `*`, but `;` and `:` are used in shaders, why we should remove them?

File: resharper/resharper-unity/src/HlslSupport/CgIncludeDirectoryTracker.cs (hunk 1 above)
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/HlslSupport/CgIncludeDirectoryTracker.cs:35]: I have changed to `IsCacheStarted`, it should be better

File: resharper/resharper-unity/src/ShaderLab/Feature/Services/TypingAssists/ShaderLabTypingAssist.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (citizenmatt) [reply resharper/resharper-unity/src/ShaderLab/Feature/Services/TypingAssists/ShaderLabTypingAssist.cs:74]: Ugh. I'm getting confused between the ShaderLab language and the injected HLSL support. I was thinking HLSL would have its own typing assist, and this was the one for the non-injected ShaderLab bits, which don't use `;` or `:` in the same way as C++

**Pull request merged**